### ISSUE TITLE
Support replicating materialized views

### ISSUE THREAD

**Issue by DuncanSands** [2ndQuadrant/pglogical#60]
Replicating tables works fine.  But when I try to add a materialized view to a replication set I get:

  select pglogical.replication_set_add_table('default','matview',True);
  ERROR:  Only tables and sequences can be added to replication set

Is there any fundamental reason why materialized views can't be replicated?  Of course you need to have a table not a materialized view on the subscriber end.

pglogical 1.2.1,postgres 9.5.4.

**Comment by PJMODOS** [comment]
Yes there are some fundamental issues in the way how postgres handles the materialized view refresh. I would recommend just replicating the refresh command via pglogical.replicate_ddl_command()

**Comment by DuncanSands** [comment]
Hi Petr,

On 11/26/2016 04:40 PM, Petr Jelinek wrote:
> Yes there are some fundamental issues in the way how postgres handles the
> materialized view refresh.

thanks for replying.  Are you referring to non-concurrent refreshes?

  I would recommend just replicating the refresh
> command via pglogical.replicate_ddl_command()

Doing as you suggest isn't feasible because the materialized view is summarizing 
terabytes of data that I definitely don't want to replicate.

Best wishes, Duncan.

**Comment by ringerc** [comment]
On 26 November 2016 at 23:51, Duncan Sands <notifications@github.com> wrote:

> Hi Petr,
>
> On 11/26/2016 04:40 PM, Petr Jelinek wrote:
> > Yes there are some fundamental issues in the way how postgres handles the
> > materialized view refresh.
>
> thanks for replying. Are you referring to non-concurrent refreshes?
>
> I would recommend just replicating the refresh
> > command via pglogical.replicate_ddl_command()
>
> Doing as you suggest isn't feasible because the materialized view is
> summarizing
> terabytes of data that I definitely don't want to replicate.
>

 pglogical.replicate_ddl_command() replicates the command to each node.

We don't replicate the data for materialized views.

-- 
 Craig Ringer                   http://www.2ndQuadrant.com/
 PostgreSQL Development, 24x7 Support, Training & Services

**Comment by DuncanSands** [comment]
Hi Craig,

> pglogical.replicate_ddl_command() replicates the command to each node.
>
> We don't replicate the data for materialized views.

that doesn't help me: in my case the materialized view summarizes about 5 
terabytes worth of tables.  I only need to replicate the materialized view, 
which is relatively small.  Replicating the original tables just to recreate the 
materialized view on the other end would be hugely more expensive.  So I would 
very much like to see pglogical get support for replicating materialized views.

I can understand that this is not a priority for you, but I'm happy to pay to 
have someone implement this.

Best wishes, Duncan.

**Comment by ringerc** [comment]
On 28 November 2016 at 18:50, Duncan Sands <notifications@github.com> wrote:

> Hi Craig,
>
> > pglogical.replicate_ddl_command() replicates the command to each node.
> >
> > We don't replicate the data for materialized views.
>
> that doesn't help me: in my case the materialized view summarizes about 5
> terabytes worth of tables. I only need to replicate the materialized view,
> which is relatively small. Replicating the original tables just to
> recreate the
> materialized view on the other end would be hugely more expensive. So I
> would
> very much like to see pglogical get support for replicating materialized
> views.
>

I see. I thought you wanted to make sure it did _not_ replicate matviews.
Which it doesn't.

Since you actually do want to replicate the data in a materialized view
refresh, your best bet at this point is probably to do "old style"
materialised views implemented on top of a normal table, refreshed with

BEGIN;
TRUNCATE ...;
INSERT INTO ... SELECT ... ;
COMMIT;

since this _will_ be replicated by pglogical.


> I can understand that this is not a priority for you, but I'm happy to pay
> to
> have someone implement this.


It's worth reaching out to info@2ndquadrant.com in that regard.

I'm pretty sure logical decoding will see the writes to the materialized
view, but like other table rewrites pglogical does not currently handle the
resulting datastream and ignores it. I expect that it would be possible to
have pglogical replicate the results of refreshing a materialized view if
one is part of a replication set by using the decoded rewrite data. This is
tentative and full of qualifications because I haven't looked into it at
all at this point.

-- 
 Craig Ringer                   http://www.2ndQuadrant.com/
 PostgreSQL Development, 24x7 Support, Training & Services
